pex/pyenv.py

```diff
@@ -0,0 +1,211 @@
+            """,
+            flags=re.VERBOSE,
+        )
+
+        @classmethod
+        def parse(cls, pyenv, binary):
+            # type: (Pyenv, str) -> Optional[Pyenv.Shim]
+            """Parses shim information from a python binary path if it looks like a pyenv shim."""
+            if os.path.dirname(binary) != os.path.join(pyenv.root, "shims"):
+                return None
+            match = cls._SHIM_REGEX.match(os.path.basename(binary))
+            if match is None:
+                return None
+            return cls(
+                pyenv=pyenv,
+                path=binary,
+                name=match.group("name"),
+                major=match.group("major"),
+                minor=match.group("minor"),
+            )
+
+        _PYENV_CPYTHON_VERSION_LEADING_CHARS = frozenset(str(digit) for digit in range(2, 10))
+
+        def select_version(self, search_dir=None):
+            # type: (Optional[str]) -> Optional[str]
+            """Reports the active shim version for the given directory or $PWD.
+
+            If the shim is not activated, returns `None`.
+            """
+            with TRACER.timed("Calculating active version for {}...".format(self), V=6):
+                active_versions = self.pyenv.active_versions(search_dir=search_dir)
+                if active_versions:
+                    binary_name = os.path.basename(self.path)
+                    if self.name == "python" and not self.major and not self.minor:
+                        for pyenv_version in active_versions:
+                            if pyenv_version[0] in self._PYENV_CPYTHON_VERSION_LEADING_CHARS:
+                                TRACER.log(
+                                    "{} has active version {}".format(self, pyenv_version), V=6
+                                )
+                                return self.pyenv.python(pyenv_version, binary_name=binary_name)
+
+                    prefix = "{name}{major}{minor}".format(
+                        name="" if self.name == "python" else self.name,
+                        major=self.major or "",
+                        minor=".{}".format(self.minor) if self.minor else "",
+                    )
+                    for pyenv_version in active_versions:
+                        if pyenv_version.startswith(prefix):
+                            TRACER.log("{} has active version {}".format(self, pyenv_version), V=6)
+                            return self.pyenv.python(pyenv_version, binary_name=binary_name)
+
+                TRACER.log("{} is not activated.".format(self), V=6)
+                return None
+
+    def as_shim(self, binary):
+        # type: (str) -> Optional[Shim]
+        """View the given binary path as a pyenv shim script if it is one."""
+        return self.Shim.parse(self, binary)
+
+    @staticmethod
+    def _read_pyenv_versions(version_file):
+        # type: (str) -> Iterator[str]
+        with open(version_file) as fp:
+            for line in fp:
```

Comment:
Why not just read the entire file into a string and split it on any whitespace?

Reply:
No need to re-open to fix.

Reply:
Yeah, that would work. I started with just returning stripped lines since thats what version files generated by `pyenv global` and `pyenv local` had and only later read the docs closer and noticed the bit about `... or any whitespace`.